x-pack/legacy/plugins/lens/public/editor_frame_plugin/embeddable/embeddable.tsx

```diff
@@ -0,0 +1,139 @@
+
+export interface LensEmbeddableConfiguration {
+  savedVis: Document;
+  editUrl: string;
+  editable: boolean;
+  indexPatterns?: StaticIndexPattern[];
+}
+
+export interface LensEmbeddableInput extends EmbeddableInput {
+  timeRange?: TimeRange;
+  query?: Query;
+  filters?: Filter[];
+}
+
+export interface LensEmbeddableOutput extends EmbeddableOutput {
+  indexPatterns?: StaticIndexPattern[];
+}
+
+export class Embeddable extends AbstractEmbeddable<LensEmbeddableInput, LensEmbeddableOutput> {
+  type = DOC_TYPE;
+
+  private expressionRenderer: ExpressionRenderer;
+  private savedVis: Document;
+  private domNode: HTMLElement | Element | undefined;
+  private subscription: Subscription;
+
+  private currentContext: {
+    timeRange?: TimeRange;
+    query?: Query;
+    filters?: Filter[];
+    lastReloadRequestTime?: number;
+  } = {};
+
+  constructor(
+    expressionRenderer: ExpressionRenderer,
+    { savedVis, editUrl, editable, indexPatterns }: LensEmbeddableConfiguration,
+    initialInput: LensEmbeddableInput,
+    parent?: IContainer
+  ) {
+    super(
+      initialInput,
+      {
+        defaultTitle: savedVis.title,
+        savedObjectId: savedVis.id!,
+        editable,
+        editUrl,
+        indexPatterns,
+      },
+      parent
+    );
+
+    this.expressionRenderer = expressionRenderer;
+    this.savedVis = savedVis;
+    this.subscription = this.getInput$().subscribe(input => this.onContainerStateChanged(input));
+    this.onContainerStateChanged(initialInput);
+  }
+
+  onContainerStateChanged(containerState: LensEmbeddableInput) {
+    const cleanedFilters = containerState.filters
+      ? containerState.filters.filter(filter => !filter.meta.disabled)
+      : undefined;
+    if (
+      !_.isEqual(containerState.timeRange, this.currentContext.timeRange) ||
+      !_.isEqual(containerState.query, this.currentContext.query) ||
+      !_.isEqual(cleanedFilters, this.currentContext.filters)
+    ) {
+      this.currentContext = {
+        timeRange: containerState.timeRange,
+        query: containerState.query,
+        lastReloadRequestTime: this.currentContext.lastReloadRequestTime,
+        filters: cleanedFilters,
+      };
+
+      if (this.domNode) {
+        this.render(this.domNode);
+      }
+    }
+  }
+
+  /**
+   *
+   * @param {HTMLElement} domNode
+   * @param {ContainerState} containerState
+   */
+  render(domNode: HTMLElement | Element) {
+    this.domNode = domNode;
+    render(
+      <ExpressionWrapper
+        ExpressionRenderer={this.expressionRenderer}
+        expression={this.savedVis.expression}
+        context={this.currentContext}
+      />,
+      domNode
+    );
+  }
+
+  destroy() {
+    super.destroy();
+    if (this.domNode) {
+      unmountComponentAtNode(this.domNode);
+    }
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+  }
+
+  reload() {
+    this.currentContext = {
+      ...this.currentContext,
+      lastReloadRequestTime: Date.now(),
```

Comment:
I would suggest using `moment` if you're planning on doing anything like date math, the regular `Date.now()` constructor is particularly tricky to test and handles time zones badly.

Reply:
This is just here to check locally whether we have to re-fetch (kind of like cache-busting), so a simple comparison is everything happens. The dashboard app uses the same thing. I also actually added the check to the `reload` function, by doing so we don't reload multiple times even if the container calls reload twice or thrice.